README.md
# Vector Math Library for Elixir

![Build Status](https://github.com/pkinney/vector_ex/actions/workflows/ci.yaml/badge.svg)
[![Hex.pm](https://img.shields.io/hexpm/v/vector.svg)](https://hex.pm/packages/vector)
[![Documentation](https://img.shields.io/badge/documentation-gray)](https://hexdocs.pm/vector)

Library of common vector functions for use in geometric or graphical calculations.

## Installation

```elixir
defp deps do
  [{:vector, "~> 1.1"}]
end
```

## Usage

**[Full Documentation](https://hexdocs.pm/vector/Vector.html)**

The `Vector` module contains several functions that take one or more vectors.
Each vector can be a 2- or 3-element tuple.  It is possible to mix two- and
three-dimensional vectors, in which case the `z` of the two-dimensional vector
will be assumed 0.

```elixir
Vector.dot({2, 3}, {1, 4}) #=> 14
Vector.cross({2, 0, -1}, {0, 3, 3}) #=> {3, -6, 6}
Vector.norm({3, -6, 6}) #=> 9
Vector.unit({2, -1}) #=> {0.894, -0.447}
Vector.add({2, 0, -1}, {0, 3, 3}) #=> {2, 3, 1}
Vector.subtract({2, 0, -1}, {1, 3}) #=> {1, -3, 1}
Vector.component({2, 3, -2}, :y) #=> 3
```

## Tests

```bash
> mix test
```

Included in the project are a suite of
[algebraic identities](https://en.wikipedia.org/wiki/Vector_algebra_relations#Addition_and_multiplication_of_vectors)
that can be run against a large generated set of vectors.


CHANGELOG.md
# Changelog

## [1.1.0] - 2023-01-09

### Changed

- Updated CI to use GitHub Actions instead of travis-ci
- Updated dependencies 


test/vector_test.exs
defmodule VectorTest do
  use ExUnit.Case, async: true

  import Vector
  doctest Vector

  test "basis vectors" do
    assert unit(basis(:x)) == basis(:x)
    assert unit(basis(:y)) == basis(:y)
    assert unit(basis(:z)) == basis(:z)

    assert dot(basis(:x), basis(:y)) == 0
    assert dot(basis(:y), basis(:z)) == 0
    assert dot(basis(:x), basis(:z)) == 0

    assert cross_norm(basis(:x), basis(:y)) == 1
    assert cross_norm(basis(:y), basis(:z)) == 1
    assert cross_norm(basis(:x), basis(:z)) == 1
  end

  test "component of a vector" do
    assert component({-1, 2}, :x) == -1
    assert component({-1, 2}, :y) == 2
    assert component({-1, 2}, :z) == 0

    assert component({-1, 2, 1}, :x) == -1
    assert component({-1, 2, 1}, :y) == 2
    assert component({-1, 2, 1}, :z) == 1
  end
end


test/vector_identities_test.exs
defmodule VectorIdentitiesTest do
  use ExUnit.Case, async: true

  import Vector

  @values [-2, 0, 1, 42]
  @vectors for(x <- @values, y <- @values, do: {x, y}) ++
             for(x <- @values, y <- @values, z <- @values, do: {x, y, z})
  @tolerance 0.00001

  defp are_equal?(r1, r2) when is_tuple(r1) and is_tuple(r2), do: equal?(r1, r2, @tolerance)
  defp are_equal?(r1, r2), do: abs(r1 - r2) <= @tolerance

  def test_identity_2(left, right) do
    for a <- @vectors, b <- @vectors, do: assert(are_equal?(left.(a, b), right.(a, b)))
  end

  def test_identity_3(left, right) do
    for a <- @vectors,
        b <- @vectors,
        c <- @vectors,
        do: assert(are_equal?(left.(a, b, c), right.(a, b, c)))
  end

  test "A + B = B + A" do
    test_identity_2(
      fn a, b -> add(a, b) end,
      fn a, b -> add(b, a) end
    )
  end

  test "A . B = B . A" do
    test_identity_2(
      fn a, b -> dot(a, b) end,
      fn a, b -> dot(b, a) end
    )
  end

  test "A x B = -B x A" do
    test_identity_2(
      fn a, b -> cross(a, b) end,
      fn a, b -> reverse(cross(b, a)) end
    )
  end

  test "(A x B) . (A x B)) = |A x B|^2 = (A . A)(B . B) - (A . B)^2" do
    test_identity_2(
      fn a, b -> dot(cross(a, b), cross(a, b)) end,
      fn a, b -> :math.pow(norm(cross(a, b)), 2) end
    )

    test_identity_2(
      fn a, b -> dot(cross(a, b), cross(a, b)) end,
      fn a, b -> dot(a, a) * dot(b, b) - :math.pow(dot(a, b), 2) end
    )
  end

  test "|^A| = |^B|" do
    f = fn v ->
      case norm(v) do
        0 -> 1
        _ -> norm(unit(v))
      end
    end

    test_identity_2(
      fn a, _ -> f.(a) end,
      fn _, b -> f.(b) end
    )
  end

  test "||A x B||^2 + (A . B)^2 = ||A||^2 ||B||^2" do
    test_identity_2(
      fn a, b -> :math.pow(norm(cross(a, b)), 2) + :math.pow(dot(a, b), 2) end,
      fn a, b -> :math.pow(norm(a), 2) * :math.pow(norm(b), 2) end
    )
  end

  test "(A + B) . C = A . C + B . C" do
    test_identity_3(
      fn a, b, c -> a |> add(b) |> dot(c) end,
      fn a, b, c -> dot(a, c) + dot(b, c) end
    )
  end

  test "(A + B) x C = A x C + B x C" do
    test_identity_3(
      fn a, b, c -> cross(add(a, b), c) end,
      fn a, b, c -> add(cross(a, c), cross(b, c)) end
    )
  end

  test "A . (B x C) = B . (C x A) = C . (A x B)" do
    test_identity_3(
      fn a, b, c -> dot(a, cross(b, c)) end,
      fn a, b, c -> dot(b, cross(c, a)) end
    )

    test_identity_3(
      fn a, b, c -> dot(b, cross(c, a)) end,
      fn a, b, c -> dot(c, cross(a, b)) end
    )
  end

  test "A x (B x C) = B (A . C) - C (A . B)" do
    test_identity_3(
      fn a, b, c -> cross(a, cross(b, c)) end,
      fn a, b, c -> subtract(multiply(b, dot(a, c)), multiply(c, dot(a, b))) end
    )
  end
end


lib/vector.ex
defmodule Vector do
  @moduledoc ~S"""
  A library of two- and three-dimensional vector operations.  All vectors
  are represented as tuples with either two or three elements.

  ## Examples
      iex> # Vector Tripple Product Identity
      ...> a = {2, 3, 1}
      ...> b = {1, 4, -2}
      ...> c = {-1, 2, 1}
      ...> Vector.equal?(
      ...>   Vector.cross(Vector.cross(a, b), c),
      ...>   Vector.subtract(Vector.multiply(b, Vector.dot(a, c)), Vector.multiply(a, Vector.dot(b, c))))
      true
  """

  @type vector :: {number, number} | {number, number, number}
  @type location :: {number, number} | {number, number, number}

  @doc ~S"""
  Returns the cross product of two vectors *A*⨯*B*

  ## Examples

      iex> Vector.cross({2, 3}, {1, 4})
      {0, 0, 5}
      iex> Vector.cross({2, 2, -1}, {1, 4, 2})
      {8, -5, 6}
      iex> Vector.cross({3, -3, 1}, {4, 9, 2})
      {-15, -2, 39}
  """
  @spec cross(vector, vector) :: vector
  def cross({x1, y1}, {x2, y2}), do: {0, 0, x1 * y2 - y1 * x2}
  def cross({x1, y1, z1}, {x2, y2}), do: cross({x1, y1, z1}, {x2, y2, 0})
  def cross({x1, y1}, {x2, y2, z2}), do: cross({x1, y1, 0}, {x2, y2, z2})

  def cross({x1, y1, z1}, {x2, y2, z2}) do
    {y1 * z2 - z1 * y2, z1 * x2 - x1 * z2, x1 * y2 - y1 * x2}
  end

  @doc ~S"""
  Returns the norm (magnitude) of the cross product of two vectors *A*⨯*B*

  ## Examples

      iex> Vector.cross_norm({2, 3}, {1, 4})
      5
      iex> Vector.cross_norm({1, 4}, {2, 2})
      6
      iex> Vector.cross_norm({2, 0, -1}, {0, 3, 3})
      9.0
      iex> Float.floor(:math.pow(Vector.cross_norm({2, 2, -1}, {1, 4, 2}), 2))
      125.0
  """
  @spec cross_norm(vector, vector) :: number
  def cross_norm({x1, y1}, {x2, y2}), do: cross_norm({x1, y1, 0}, {x2, y2, 0})

  def cross_norm({x1, y1, z1}, {x2, y2, z2}) do
    norm(cross({x1, y1, z1}, {x2, y2, z2}))
  end

  @doc ~S"""
  Returns the dot product of two vectors *A*⨯*B*

  ## Examples

      iex> Vector.dot({2, 3}, {1, 4})
      14
      iex> Vector.dot({1, 4}, {2, 2})
      10
      iex> Vector.dot({2, 0, -1}, {0, 3, 3})
      -3
  """
  @spec dot(vector, vector) :: number
  def dot({x1, y1}, {x2, y2}), do: dot({x1, y1, 0}, {x2, y2, 0})
  def dot({x1, y1, z1}, {x2, y2}), do: dot({x1, y1, z1}, {x2, y2, 0})
  def dot({x1, y1}, {x2, y2, z2}), do: dot({x1, y1, 0}, {x2, y2, z2})

  def dot({x1, y1, z1}, {x2, y2, z2}) do
    x1 * x2 + y1 * y2 + z1 * z2
  end

  @doc ~S"""
  Returns the norm (magnitude) of a vector

  ## Examples

      iex> Vector.norm({3, 4})
      5.0
      iex> Vector.norm({-1, 0})
      1
      iex> Vector.norm({0, -2, 0})
      2
  """
  @spec norm(vector) :: number
  def norm({x, y}), do: norm({x, y, 0})
  def norm({0, 0, z}), do: abs(z)
  def norm({x, 0, 0}), do: abs(x)
  def norm({0, y, 0}), do: abs(y)
  def norm({x, y, z}), do: :math.sqrt(norm_squared({x, y, z}))

  @doc ~S"""
  Returns the square of the norm norm (magnitude) of a vector

  ## Examples

      iex> Vector.norm_squared({3, 4})
      25
      iex> Vector.norm_squared({1, 0})
      1
      iex> Vector.norm_squared({2, 0, -1})
      5
      iex> Vector.norm_squared({-2, 3, 1})
      14
  """
  @spec norm_squared(vector) :: number
  def norm_squared({x, y}), do: norm_squared({x, y, 0})

  def norm_squared({x, y, z}) do
    x * x + y * y + z * z
  end

  @doc ~S"""
  Returns the unit vector parallel ot the given vector.
  This will raise an `ArithmeticError` if a zero-magnitude vector is given.
  Use `unit_safe` if there is a chance that a zero-magnitude vector
  will be sent.

  ## Examples

      iex> Vector.unit({3, 4})
      {0.6, 0.8}
      iex> Vector.unit({8, 0, 6})
      {0.8, 0.0, 0.6}
      iex> Vector.unit({-2, 0, 0})
      {-1.0, 0.0, 0.0}
      iex> Vector.unit({0, 0, 0})
      ** (ArithmeticError) bad argument in arithmetic expression
  """
  @spec unit(vector) :: vector
  def unit({x, 0, 0}), do: {x / abs(x), 0.0, 0.0}
  def unit({0, y, 0}), do: {0.0, y / abs(y), 0.0}
  def unit({0, 0, z}), do: {0.0, 0.0, z / abs(z)}
  def unit(v), do: divide(v, norm(v))

  @doc ~S"""
  Returns the unit vector parallel ot the given vector, but will handle
  the vectors `{0, 0}`  and `{0, 0, 0}` by returning the same vector

  ## Examples

      iex> Vector.unit_safe({3, 4})
      {0.6, 0.8}
      iex> Vector.unit_safe({0, 0})
      {0, 0}
      iex> Vector.unit_safe({0, 0, 0})
      {0, 0, 0}
  """
  @spec unit_safe(vector) :: vector
  def unit_safe({0, 0}), do: {0, 0}
  def unit_safe({0, 0, 0}), do: {0, 0, 0}
  def unit_safe(v), do: unit(v)

  @doc ~S"""
  Reverses a vector

  ## Examples

      iex> Vector.reverse({3, -4})
      {-3, 4}
      iex> Vector.reverse({-2, 0, 5})
      {2, 0, -5}
      iex> Vector.cross_norm({-2, 3, 5}, Vector.reverse({-2, 3, 5}))
      0
  """
  @spec reverse(vector) :: vector
  def reverse({x, y}), do: {-x, -y}
  def reverse({x, y, z}), do: {-x, -y, -z}

  @doc ~S"""
  Adds two vectors

  ## Examples

      iex> Vector.add({3, -4}, {2, 1})
      {5,-3}
      iex> Vector.add({-2, 0, 5}, {0, 0, 0})
      {-2, 0, 5}
      iex> Vector.add({2, 1, -2}, Vector.reverse({2, 1, -2}))
      {0, 0, 0}
  """
  @spec add(vector, vector) :: vector
  def add({x1, y1}, {x2, y2}), do: {x1 + x2, y1 + y2}
  def add({x1, y1, z1}, {x2, y2}), do: add({x1, y1, z1}, {x2, y2, 0})
  def add({x1, y1}, {x2, y2, z2}), do: add({x1, y1, 0}, {x2, y2, z2})
  def add({x1, y1, z1}, {x2, y2, z2}), do: {x1 + x2, y1 + y2, z1 + z2}

  @doc ~S"""
  Subtract vector *B* from vector *A*.  Equivalent to `Vector.add(A, Vector.revers(B))`

  ## Examples

      iex> Vector.subtract({3, -4}, {2, 1})
      {1,-5}
      iex> Vector.subtract({-2, 0, 5}, {-3, 1, 2})
      {1, -1, 3}
  """
  @spec subtract(vector, vector) :: vector
  def subtract(a, b), do: add(a, reverse(b))

  @doc ~S"""
  Multiply a vector by scalar value `s`

  ## Examples

      iex> Vector.multiply({3, -4}, 2.5)
      {7.5, -10.0}
      iex> Vector.multiply({-2, 0, 5}, -2)
      {4, 0, -10}
  """
  @spec multiply(vector, number) :: vector
  def multiply({x, y}, s), do: {x * s, y * s}
  def multiply({x, y, z}, s), do: {x * s, y * s, z * s}

  @doc ~S"""
  Divide a vector by scalar value `s`

  ## Examples

      iex> Vector.divide({3, -4}, 2.5)
      {1.2, -1.6}
      iex> Vector.divide({-2, 0, 5}, -2)
      {1.0, 0.0, -2.5}
  """
  @spec divide(vector, number) :: vector
  def divide({x, y}, s), do: {x / s, y / s}
  def divide({x, y, z}, s), do: {x / s, y / s, z / s}

  @doc ~S"""
  Returns a new coordinate by projecting a given length `distance` from
  coordinate `start` along `vector`

  ## Examples

      iex> Vector.project({3, -4}, {-1, 1}, 4)
      {1.4, -2.2}
      iex> Vector.project({-6, 0, 8}, {1, -2, 0.4}, 2.5)
      {-0.5, -2.0, 2.4}
      iex> Vector.project({-2, 1, 3}, {0, 0, 0}, 2.5) |> Vector.norm()
      2.5
  """
  @spec project(vector, location, number) :: location
  def project(vector, start, distance) do
    vector
    |> unit()
    |> multiply(distance)
    |> add(start)
  end

  @doc ~S"""
  Compares two vectors for equality, with an optional tolerance

  ## Examples

      iex> Vector.equal?({3, -4}, {3, -4})
      true
      iex> Vector.equal?({3, -4}, {3.0001, -3.9999})
      false
      iex> Vector.equal?({3, -4}, {3.0001, -3.9999}, 0.001)
      true
      iex> Vector.equal?({3, -4, 1}, {3.0001, -3.9999, 1.0}, 0.001)
      true
  """
  @spec equal?(vector, vector, number) :: boolean
  def equal?(a, b, tolerance \\ 0.0) do
    norm_squared(subtract(a, b)) <= tolerance * tolerance
  end

  @doc ~S"""
  Returns the scalar component for the axis given

  ## Examples

      iex> Vector.component({3, -4}, :y)
      -4
      iex> Vector.component({-6, 0, 8}, :z)
      8
      iex> Vector.component({1, -2}, :z)
      0
      iex> Vector.component(Vector.basis(:x), :z)
      0
  """
  @spec component(vector, atom) :: number
  def component({x, _}, :x), do: x
  def component({_, y}, :y), do: y
  def component({_, _}, :z), do: 0
  def component({x, _, _}, :x), do: x
  def component({_, y, _}, :y), do: y
  def component({_, _, z}, :z), do: z

  @doc ~S"""
  Returns the basis vector for the given axis

  ## Examples

      iex> Vector.basis(:x)
      {1, 0, 0}
      iex> Vector.basis(:y)
      {0, 1, 0}
      iex> Vector.component(Vector.basis(:y), :y)
      1
  """
  @spec basis(atom) :: vector
  def basis(:x), do: {1, 0, 0}
  def basis(:y), do: {0, 1, 0}
  def basis(:z), do: {0, 0, 1}
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
import Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :vector, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:vector, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


bench/vector_bench.exs
defmodule VectorBench do
  use Benchfella
  import Vector

  @values [-1, 0, 1, 2]
  @vectors (for x <- @values, y <- @values, do: {x, y}) ++ (for x <- @values, y <- @values, z <- @values, do: {x, y, z})
  @zero_2 {0, 0}
  @zero_3 {0, 0, 0}

  def run_vector_set_2(func) do
    for a <- @vectors, b <- @vectors, do: func.(a, b)
  end

  def run_vector_set_3(func) do
    for a <- @vectors, b <- @vectors, c <- @vectors,  do: func.(a, b, c)
  end

  bench "zero-vector cross 2D" do
    Vector.cross(@zero_2, @zero_2)
  end

  bench "zero-vector cross 3D" do
    Vector.cross(@zero_3, @zero_3)
  end

  bench "unit vector" do
    run_vector_set_2 fn a, b-> cross(unit_safe(a), unit_safe(b)) end
  end

  bench "A x B" do
    run_vector_set_2 fn a, b -> cross(a, b) end
  end

  bench "A . B" do
    run_vector_set_2 fn a, b -> dot(a, b) end
  end

  bench "A . (B x C)" do
    run_vector_set_3 fn a, b, c -> cross(a, cross(b, c)) end
  end

  bench "A x (B x C)" do
    run_vector_set_3 fn a, b, c -> dot(a, cross(b, c)) end
  end

  bench "B (A . C) - C (A . B)" do
    run_vector_set_3 fn a, b, c -> subtract(multiply(b, dot(a, c)), multiply(c, dot(a, b))) end
  end
end


mix.exs
defmodule Vector.Mixfile do
  use Mix.Project

  def project do
    [
      app: :vector,
      version: "1.1.0",
      elixir: "~> 1.2",
      description: description(),
      package: package(),
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [coveralls: :test],
      dialyzer: [plt_add_apps: [:poison, :mix]],
      deps: deps(),
      aliases: aliases()
    ]
  end

  # Configuration for the OTP application
  #
  # Type "mix help compile.app" for more information
  def application do
    [applications: [:logger]]
  end

  defp deps do
    [
      {:earmark, "~> 1.0", only: :dev},
      {:dialyxir, "~> 1.0", only: [:dev], runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:credo, "~> 1.5", only: :dev, runtime: false},
      {:benchfella, "~> 0.3", only: :dev},
      {:excoveralls, "~> 0.4", only: :test}
    ]
  end

  defp description do
    """
    Library of common vector functions for use in geometric or graphical calculations.
    """
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README*"],
      maintainers: ["Powell Kinney"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/pkinney/vector_ex",
        "Docs" => "https://hexdocs.pm/vector/Vector.html"
      }
    ]
  end

  defp aliases do
    [
      validate: [
        "clean",
        "compile --warnings-as-error",
        "format --check-formatted",
        "credo",
        "dialyzer"
      ]
    ]
  end
end
